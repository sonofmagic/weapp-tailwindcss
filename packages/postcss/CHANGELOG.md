# @weapp-tailwindcss/postcss

## 1.0.0-alpha.1

### Major Changes

- [`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: monorepo changes

### Patch Changes

- Updated dependencies [[`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875)]:
  - @weapp-tailwindcss/shared@1.0.0-alpha.1

## 0.0.1-alpha.0

### Patch Changes

- Updated dependencies [[`d50d95d`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d50d95d04e1c6b7c3ff32acc0d9894d3c0f06d22)]:
  - @weapp-tailwindcss/shared@0.0.1-alpha.0
