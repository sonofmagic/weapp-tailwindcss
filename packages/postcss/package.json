{
  "name": "@weapp-tailwindcss/postcss",
  "version": "1.0.0-alpha.2",
  "description": "@weapp-tailwindcss/postcss",
  "author": "ice breaker <1324318532@qq.com>",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sonofmagic/monorepo-template.git",
    "directory": "packages/bar"
  },
  "bugs": {
    "url": "https://github.com/sonofmagic/monorepo-template/issues"
  },
  "keywords": [],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    },
    "./types": {
      "types": "./dist/types.d.ts",
      "import": "./dist/types.mjs",
      "require": "./dist/types.js"
    }
  },
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "dev": "tsup --watch --sourcemap",
    "build": "tsup",
    "test": "vitest run",
    "test:dev": "vitest",
    "release": "pnpm publish",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix"
  },
  "publishConfig": {
    "exports": {
      ".": {
        "types": "./dist/index.d.ts",
        "import": "./dist/index.mjs",
        "require": "./dist/index.js"
      },
      "./types": {
        "types": "./dist/types.d.ts",
        "import": "./dist/types.mjs",
        "require": "./dist/types.js"
      }
    },
    "main": "./dist/index.js",
    "module": "./dist/index.mjs",
    "types": "./dist/index.d.ts"
  },
  "dependencies": {
    "@csstools/postcss-is-pseudo-class": "^5.0.1",
    "@weapp-core/escape": "~3.0.2",
    "@weapp-tailwindcss/shared": "workspace:*",
    "postcss": "~8.4.49",
    "postcss-rem-to-responsive-pixel": "~6.0.2",
    "postcss-selector-parser": "~7.0.0"
  },
  "devDependencies": {
    "@weapp-tailwindcss/mangle": "workspace:*"
  }
}
