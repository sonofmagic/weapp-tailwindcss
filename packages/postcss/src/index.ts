// 统一导出入口，供外部调用端按需引用核心能力
export * from './handler'
export {
  createStylePipeline,
  type PipelineNodeContext,
  type PipelineNodeCursor,
  type PipelineStage,
  type ResolvedPipelineNode,
  type StyleProcessingPipeline,
} from './pipeline'
export { createInjectPreflight } from './preflight'
export { internalCssSelectorReplacer } from './shared'
export * from './types'
