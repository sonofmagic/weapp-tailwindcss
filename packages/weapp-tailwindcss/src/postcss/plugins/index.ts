import type { AcceptedPlugin } from 'postcss'
// @ts-expect-error
import postcssIsPseudoClass from '@csstools/postcss-is-pseudo-class'
import postcssRem2rpx from 'postcss-rem-to-responsive-pixel'
import type { IStyleHandlerOptions } from '../../types'
import { postcssWeappTailwindcssPrePlugin } from './pre'
import { postcssWeappTailwindcssPostPlugin } from './post'
import { createContext } from './ctx'

export function getPlugins(options: IStyleHandlerOptions) {
  const ctx = createContext()
  options.ctx = ctx
  const plugins: AcceptedPlugin[] = [
    ...(options.postcssOptions?.plugins ?? []),
    postcssWeappTailwindcssPrePlugin(options),
    postcssIsPseudoClass({
      specificityMatchingName: 'weapp-tw-ig',
    }),
  ]
  if (options.rem2rpx) {
    plugins.push(
      postcssRem2rpx(
        typeof options.rem2rpx === 'object'
          ? options.rem2rpx
          : {
              rootValue: 32,
              propList: ['*'],
              transformUnit: 'rpx',
            },
      ),
    )
  }
  plugins.push(postcssWeappTailwindcssPostPlugin(options))
  return plugins
}

export { postcssWeappTailwindcssPostPlugin } from './post'
export { postcssWeappTailwindcssPrePlugin } from './pre'
// @ts-expect-error
export { default as postcssIsPseudoClass } from '@csstools/postcss-is-pseudo-class'
export { default as postcssRem2rpx } from 'postcss-rem-to-responsive-pixel'
