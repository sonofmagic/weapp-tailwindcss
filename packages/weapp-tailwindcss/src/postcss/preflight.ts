import type { CssPreflightOptions, IPropValue } from '../types'

export type InjectPreflight = () => IPropValue[]

export function createInjectPreflight(options?: CssPreflightOptions): InjectPreflight {
  const result: IPropValue[] = []
  // if options false ,do no thing
  if (options && typeof options === 'object') {
    const entries = Object.entries(options)
    for (const [prop, value] of entries) {
      if (value !== false) {
        result.push({
          prop,
          value: value.toString(),
        })
      }
    }
  }

  return () => {
    return result
  }
}
