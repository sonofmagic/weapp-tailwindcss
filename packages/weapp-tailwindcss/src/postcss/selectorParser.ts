import selectorParser from 'postcss-selector-parser'
import type { SyncProcessor } from 'postcss-selector-parser'
import type { Rule } from 'postcss'
import { composeIsPseudo, internalCssSelectorReplacer } from './shared'
import type { IStyleHandlerOptions } from '@/types'

const createRuleTransform = (rule: Rule, options: IStyleHandlerOptions) => {
  const { escapeMap, mangleContext, cssSelectorReplacement, cssRemoveHoverPseudoClass } = options

  const transform: SyncProcessor = (selectors) => {
    selectors.walk((selector) => {
      // do something with the selector
      // node.selector.replace(/\*/g, 'view')
      if (selector.type === 'universal' && cssSelectorReplacement && cssSelectorReplacement.universal) {
        selector.value = composeIsPseudo(cssSelectorReplacement.universal)
      }

      if (cssRemoveHoverPseudoClass && selector.type === 'selector') {
        const node = selector.nodes.find((x) => x.type === 'pseudo' && x.value === ':hover')
        node && selector.remove()
      }

      if (
        selector.type === 'pseudo' &&
        selector.value === ':root' &&
        cssSelectorReplacement &&
        cssSelectorReplacement.root
      ) {
        selector.value = composeIsPseudo(cssSelectorReplacement.root)
      }

      if (selector.type === 'class') {
        selector.value = internalCssSelectorReplacer(selector.value, {
          escapeMap,
          mangleContext
        })
      }
    })
    if (selectors.length === 0) {
      rule.remove()
    }
  }
  return transform
}

const getRuleTransformer = (rule: Rule, options: IStyleHandlerOptions) => {
  return selectorParser(createRuleTransform(rule, options))
}

export const ruleTransformSync = (rule: Rule, options: IStyleHandlerOptions) => {
  const transformer = getRuleTransformer(rule, options)

  return transformer.transformSync(rule, {
    lossless: false,
    updateSelector: true
  })
}

export function isOnlyBeforeAndAfterPseudoElement(node: Rule) {
  let b = false
  let a = false

  selectorParser((selectors) => {
    selectors.walkPseudos((s) => {
      if (s.parent?.length === 1) {
        if (/^:?:before$/.test(s.value)) {
          b = true
        }
        if (/^:?:after$/.test(s.value)) {
          a = true
        }
      }
    })
  }).astSync(node)

  return b && a
}

export const fallbackRemove = selectorParser((selectors) => {
  selectors.walk((selector) => {
    if (selector.type === 'universal') {
      selector.parent?.remove()
    }
    if (selector.type === 'pseudo' && selector.value === ':is') {
      selector.parent?.remove()
    }
  })
})
