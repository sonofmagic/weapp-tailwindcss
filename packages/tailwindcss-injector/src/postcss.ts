import type { Input, PluginCreator } from 'postcss'
import type { Config } from 'tailwindcss'
import type { Options } from './types'
import postcss from 'postcss'
import tailwindcss from 'tailwindcss'
import { loadConfig } from 'tailwindcss-config'
import { getConfig } from './config'
import { postcssPlugin } from './constants'
import { regExpTest, removeFileExtension } from './utils'
import { getDepFiles } from './wxml'

const configCache = new Map<string, Promise<Config | undefined>>()

async function loadTailwindConfigOnce(cwd: string, configPath: string) {
  const cacheKey = `${cwd}::${configPath}`
  let cached = configCache.get(cacheKey)
  if (!cached) {
    cached = loadConfig({
      cwd,
      config: configPath,
    }).then(result => result?.config as Config | undefined)
    configCache.set(cacheKey, cached)
  }
  return cached
}

function cloneConfig(config: Config | undefined): Config | undefined {
  if (!config) {
    return config
  }

  return {
    ...config,
  }
}

async function resolveTailwindConfig(config: Options['config'], cwd: string, sourceInput: Input | undefined) {
  if (!config) {
    return undefined
  }

  const resolved = typeof config === 'function' ? config(sourceInput) : config

  if (typeof resolved === 'string') {
    const loaded = await loadTailwindConfigOnce(cwd, resolved)
    return cloneConfig(loaded)
  }

  return cloneConfig(resolved as Config)
}

// function isObject(obj: any): obj is object {
//   return typeof obj === 'object' && obj !== null
// }
export type {
  Options,
}

const creator: PluginCreator<Partial<Options>> = (options) => {
  const { config, filter, directiveParams, cwd, extensions, insertAfterAtRulesNames, insertAfterComments } = getConfig(options)

  const extensionsGlob = `{${extensions.join(',')}}`

  return {
    postcssPlugin,
    plugins: [
      {
        postcssPlugin: `${postcssPlugin}:post`,
        async Once(root, helpers) {
          const sourceInput = root.source?.input
          if (filter(sourceInput)) {
            const nodes = root.nodes ?? []
            let atruleAnchorIndex = -1
            let commentAnchorIndex = -1
            const directivePresence = new Set<string>()

            nodes.forEach((node, index) => {
              if (node.type === 'atrule') {
                const nameMatches = insertAfterAtRulesNames.includes(node.name)
                if (nameMatches) {
                  atruleAnchorIndex = Math.max(atruleAnchorIndex, index)
                }
                if (typeof node.params === 'string') {
                  directivePresence.add(node.params)
                }
              }
              else if (node.type === 'comment' && regExpTest(insertAfterComments, node.text)) {
                commentAnchorIndex = Math.max(commentAnchorIndex, index)
              }
            })

            const anchorIndex = Math.max(atruleAnchorIndex, commentAnchorIndex)
            const anchorNode = anchorIndex >= 0 ? nodes[anchorIndex] : undefined
            let lastInserted = anchorNode

            for (const params of directiveParams) {
              if (!directivePresence.has(params)) {
                const atRule = helpers.atRule({ name: 'tailwind', params })
                if (lastInserted) {
                  root.insertAfter(lastInserted, atRule)
                }
                else {
                  root.prepend(atRule)
                }
                lastInserted = atRule
                directivePresence.add(params)
              }
            }

            if (sourceInput?.file) {
              const tailwindcssConfig = (await resolveTailwindConfig(config, cwd, sourceInput)) ?? { content: [] }

              const basename = removeFileExtension(sourceInput.file)
              const contentEntries = [`${basename}.${extensionsGlob}`]

              const deps = await getDepFiles(`${basename}.wxml`)
              for (const dep of deps) {
                contentEntries.push(`${removeFileExtension(dep)}.${extensionsGlob}`)
              }

              const uniqueEntries = Array.from(new Set(contentEntries))
              const { content: existingContent } = tailwindcssConfig as { content?: Config['content'] }

              if (!existingContent || Array.isArray(existingContent)) {
                tailwindcssConfig.content = uniqueEntries
              }
              else if (typeof existingContent === 'object' && existingContent !== null) {
                const normalizedContent = { ...existingContent }
                const currentFiles = normalizedContent.files
                const normalizedFiles = Array.isArray(currentFiles)
                  ? currentFiles
                  : typeof currentFiles === 'string'
                    ? [currentFiles]
                    : []
                normalizedContent.files = Array.from(new Set([...normalizedFiles, ...uniqueEntries]))
                tailwindcssConfig.content = normalizedContent
              }
              else {
                tailwindcssConfig.content = uniqueEntries
              }

              await postcss([
                tailwindcss(tailwindcssConfig),
              ]).process(root, {
                from: sourceInput.file,
              }).async()
            }
          }
        },
      },
    ],
  }
}
creator.postcss = true

export default creator
