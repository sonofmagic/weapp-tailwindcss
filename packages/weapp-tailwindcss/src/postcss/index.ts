import postcss from 'postcss'
import type { IStyleHandlerOptions } from '../types'
import { defuOverrideArray } from '../utils'
import { getPlugins } from './plugins'

export function styleHandler(rawSource: string, options: IStyleHandlerOptions) {
  return postcss(getPlugins(options))
    .process(
      rawSource,
      options.postcssOptions?.options ?? {
        from: undefined,
      },
    )
    .async()
}

export function createStyleHandler(options: Partial<IStyleHandlerOptions>) {
  return (rawSource: string, opt?: Partial<IStyleHandlerOptions>) => {
    return styleHandler(
      rawSource,
      defuOverrideArray<IStyleHandlerOptions, Partial<IStyleHandlerOptions>[]>(opt as IStyleHandlerOptions, options),
    )
  }
}
