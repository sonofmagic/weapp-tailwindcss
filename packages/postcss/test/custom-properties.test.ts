import postcss from 'postcss'
import postcssCustomProperties from 'postcss-custom-properties'

describe('custom-properties', () => {
  const baseCss = `:root {
--color-blue-dark: rgb(0, 61, 184);
--color-blue-light: rgb(0, 217, 255);
--color-pink: rgb(255, 192, 211);
--text-color: var(--color-pink);
}

.element {
/* custom props */
--border-color: var(--color-blue-light);
/* props */
border: 1px solid var(--border-color);
color: var(--text-color);
}

.element--dark {
--border-color: var(--color-blue-dark);
}`
  it('should case 0', () => {
    const { css } = postcss([
      postcssCustomProperties({
        preserve: true,
      }),
    ]).process(baseCss, {
      from: undefined,
    })

    expect(css).toMatchSnapshot()
  })

  it('should case 1', () => {
    const HAS_VAR_FUNCTION_REGEX = /\bvar\(/i
    const { css } = postcss([
      // {

      // } satisfies postcss.Processor
      {
        postcssPlugin: 'postcss-custom-properties',
        prepare(css) {
          // https://github.com/csstools/postcss-plugins/blob/main/plugins/postcss-custom-properties/src/index.ts
          const ppp = postcssCustomProperties({
            preserve: true,
          }) as postcss.Plugin
          const hooks = ppp.prepare?.(css)
          return {
            Once: hooks?.Once,
            Declaration(decl) {
              if (!HAS_VAR_FUNCTION_REGEX.test(decl.value)) {
                return
              }
              hooks
                // @ts-ignore
                ?.Declaration
                ?.(decl)
            },
          }
        },
      },

    ]).process(baseCss, {
      from: undefined,
    })

    expect(css).toMatchSnapshot()
  })
})
