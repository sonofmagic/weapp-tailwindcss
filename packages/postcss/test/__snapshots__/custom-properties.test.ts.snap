// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom-properties > should case 0 1`] = `
":root {
--color-blue-dark: rgb(0, 61, 184);
--color-blue-light: rgb(0, 217, 255);
--color-pink: rgb(255, 192, 211);
--text-color: var(--color-pink);
}

.element {
/* custom props */
--border-color: var(--color-blue-light);
/* props */
border: 1px solid rgb(0, 217, 255);
border: 1px solid var(--border-color);
color: rgb(255, 192, 211);
color: var(--text-color);
}

.element--dark {
--border-color: var(--color-blue-dark);
}"
`;

exports[`custom-properties > should case 1 1`] = `
":root {
--color-blue-dark: rgb(0, 61, 184);
--color-blue-light: rgb(0, 217, 255);
--color-pink: rgb(255, 192, 211);
--text-color: var(--color-pink);
}

.element {
/* custom props */
--border-color: var(--color-blue-light);
/* props */
border: 1px solid rgb(0, 217, 255);
border: 1px solid var(--border-color);
color: rgb(255, 192, 211);
color: var(--text-color);
}

.element--dark {
--border-color: var(--color-blue-dark);
}"
`;
